File: MapA7/MapA7/Model/Statements/ReadFile.cs
using MapA7.Model.Exceptions;
using MapA7.Model.Expressions;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Threading.Tasks;

namespace MapA7.Model.Statements
{
    public class ReadFile:IStatement
    {
        private IExpression expr;
        private String var;

        public ReadFile(IExpression expr, String var)
        {
            this.expr = expr;
            this.var = var;
        }

        public ProgramState Execute(ProgramState ps)
        {
            SymbolTable symTable = ps.GetSymbolTable();
            FileTable fileTable = ps.GetFileTalbe();
            int fd = expr.Evaluate(symTable);
            StreamReader sr = fileTable.GetValue(fd);
            if (sr == null) throw new CustomException("The file isn't opened!.");
            try
            {
                int value = 0;
                String line = sr.ReadLine();
                if (line != null)
                    value = System.Convert.ToInt32(line);
                symTable.Put(var, value);
            }
            catch (IOException)
            {
                throw new CustomException("Reading from the file resulted in an error.");
            }
            return ps;
        }

        public String Tostring()
        {
            return "ReadFile(" + expr.Tostring() + "," + var + ")";
        }
    }
}

File: MapGUI3/src/Model/Statements/readFile.java
package Model.Statements;

import Model.FileTable;
import Model.Heap;
import Model.ProgramState;
import Model.SymbolTable;
import Model.expressions.Expression;
import java.io.BufferedReader;
import java.io.IOException;

public class readFile implements IStatement {
    private Expression id;
    private String var_name;

    public readFile(Expression id, String var_name) {
        this.id = id;
        this.var_name = var_name;
    }

    public ProgramState execute(ProgramState ps){
        FileTable ft=ps.getFileTable();
        SymbolTable<String,Integer> symTable = ps.getSymTable();
        Heap ph=ps.getHeap();
        int res=id.evaluate(symTable,ph);
        BufferedReader buff = ft.getBufferReader(res);
        String line = null;
        try {
            line = buff.readLine();
        } catch (IOException e) {
            e.printStackTrace();
        }
        int rez;
        if (line == null) {
            rez = 0;

        } else {
            rez = Integer.parseInt(line);
        }
        symTable.put(var_name,rez);
        return ps;
    }

    public String toString(){
        return "readFile("+id.toString()+','+var_name+')';
    }
}
